Add property test for `ErrBalanceTxUnableToCreateInput`.

jonathanknowles opened the pull request:
## Issue

Follow-on from #4379

## Description

This PR adds a test for the `ErrBalanceTxUnableToCreateInput` failure condition.

In addition, this PR also:
- adds a `BalanceTxArgs` record type to hold arguments for the `balanceTx` function.
- adds a pair of combinators for `BalanceTxArgs`:
    - `Success`
      for arguments that will always produce a balanced transaction.
    - `SuccessOrFailure`
      for arguments that will sometimes produce a balanced transaction but sometimes result in failure.
- uses these combinators to simplify the arguments for `balanceTx` properties.

jonathanknowles committed:
Add property test for `ErrBalanceTxUnableToCreateInput`.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -1323,6 +1326,48 @@ prop_balanceTransactionExistingTotalCollateral
+    & cover 1
+        (isRight r0)
+        "isRight r0"
+    & checkCoverage
```
If the `Left` case is uninteresting and we take care to ensure some `Right` coverage, then wouldn't using `==>` be both easier and clearer? 🤔 

E.g.:
```haskell
isRight r0 ==> r1 === Left ErrBalanceTxUnableToCreateInput
```

jonathanknowles replied:
> If the `Left` case is uninteresting and we take care to ensure some `Right` coverage, then wouldn't using `==>` be both easier and clearer? 🤔
> 
> E.g.:
> 
> ```haskell
> isRight r0 ==> r1 === Left ErrBalanceTxUnableToCreateInput
> ```

Unfortunately, using the `==>` combinator is impractical here, as on average only a very small number of cases result in a successfully balanced transaction (around 2%). This leads to test flakiness, as QuickCheck often gives up if the LHS of `==>` is repeatedly unsatisfied.

To solve this, I've pushed some further commits to this PR that introduce a `BalanceTxArgs` record type with a `Success` combinator. The `Arbitrary` instance for `Success BalanceTxArgs` guarantees to always generate arguments that lead to balanced transactions.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -1323,6 +1326,48 @@ prop_balanceTransactionExistingTotalCollateral
+  where
+    r0 = balanceTx w pp tt seed ptx
+    r1 = balanceTx
+        (eraseWalletUTxOSet w) pp tt seed (erasePartialTxInputList ptx)
```
Style topic: why not inline `r0` and `r1`, potentially factoring out `balanceTx' = balanceTx pp tt seed` modulo argument order? Maybe we can briefly discuss tomorrow

jonathanknowles replied:
This was resolved by f2cde3177d514381add96a5a4672e18d64182edd.

jonathanknowles committed:
Add `BalanceTxArgs` record type.

jonathanknowles committed:
Add `SuccessOrFailure` combinator.

jonathanknowles committed:
Use `BalanceTxArgs` in `prop_balanceTransactionExistingTotalCollateral`.

We use the `SuccessOrFailure` combinator.

jonathanknowles committed:
Use `BalanceTxArgs` in `prop_balanceTransactionExistingReturnCollateral`.

We use the `SuccessOrFailure` combinator.

jonathanknowles committed:
Use `BalanceTxArgs` in `prop_balanceTransactionValid`.

We use the `SuccessOrFailure` combinator.

jonathanknowles committed:
Add `Success` combinator.

jonathanknowles committed:
Use `BalanceTxArgs` in `prop_balanceTransactionUnableToCreateInput`.

We use the `Success` combinator.

jonathanknowles committed:
Add `protocolParams` field to `BalanceTxArgs`.

jonathanknowles committed:
Use `protocolParams` field in property tests for `balanceTx`.

jonathanknowles committed:
Inline definition of `ledgerPParams`.

jonathanknowles committed:
Add `timeTranslation` field to `BalanceTxArgs`.

jonathanknowles committed:
Use `timeTranslation` field in properties for `balanceTx`.

jonathanknowles committed:
Add convenience function `applyBalanceTxArgs`.

jonathanknowles committed:
Use `applyBalanceTxArgs` to simplify `genBalanceTxArgsForSuccess`.

jonathanknowles committed:
Rename `shrinkBalanceTxArgs`.

This shrinker can either produce arguments that lead to success or it
can produce arguments that lead to failure.

So we rename it to `shrinkBalanceTxArgsForSuccessOrFailure` for
consistency.

jonathanknowles committed:
Add `shrinkBalanceTxArgsForSuccess`.

jonathanknowles committed:
Use `shrinkBalanceTxArgsForSuccess` to fix `Arbitrary` instance.

jonathanknowles commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -2007,6 +2065,97 @@ prop_splitSignedValue_mergeSignedValue (MixedSign v) =
+    -> [BalanceTxArgs era]
+shrinkBalanceTxArgsForSuccess
+    = filter (isRight . applyBalanceTxArgs)
+    . shrinkBalanceTxArgsForSuccess
```
```suggestion
    . shrinkBalanceTxArgsForSuccessOrFailure
```

jonathanknowles committed:
Reorder fields of `BalanceTxArgs`.

To match the actual order of arguments for `balanceTx`.

Anviking commented on lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -496,6 +510,9 @@ spec_balanceTransaction = describe "balanceTransaction" $ do
     it "doesn't balance transactions with existing 'returnCollateral'"
         $ property prop_balanceTransactionExistingReturnCollateral
 
+    it "does not balance transactions if no inputs can be created"
```
I think this description may read slightly over-simplified, when the test is about when both the utxo set and existing inputs in the tx are empty

jonathanknowles replied:
I agree the description is non-ideal! Though I'm not sure how to improve it and still keep it concise...

Anviking reviewed (approved):
I'm torn on `...Args` data types but 👍
